Change Item Hierarchy

querdenker2k opened the pull request:
Changed the Item hierarchy and removed the generation of the specific Items. Just the JRuleItems will be generated.
A XXXGroupItem will extend the XXXItem.

querdenker2k commented:
@seaside1 Important things i've done
- I removed the XYValue and RGBValue, just HSB exist. I could add RGB again if required
- Created InternalItems for every ItemType
- Added a lot of tests for the Items
- New JRuleItemName Class for referencing the Itemnames for the rule conditions. Not possible with JRuleItems anymore, because there is no static value anymore. First I tried to create an enum for this, but failed, because i don't see any option for testing

querdenker2k commented:
Using this in my productive environment now.
@seaside1 I know this is again a huge mr. What do you need for aprooving?

seime commented:
Thanks @querdenker2k . I started looking around yesterday, but as this PR is huge I would need some more time to update my rules and look at the code.

querdenker2k commented:
But i am having problems getting the integration test stable

seaside1 commented:
Sorry for not commenting earlier. Will also start to review and test it out.

querdenker2k commented:
the splitted loglines fix for the integration test seems to work now

querdenker2k commented:
There seems to be a problem sending String commands (in the right format) to any item-type.
I will change the state types in the JRuleItemEvent to the JRuleValue types 
and allow to send any kind of value to any item. Why?
I have rules like this
```
    @JRuleName("light proxy to proxy")
    @JRuleWhenItemReceivedCommand(item = JRuleItemName.G_Light_Proxy, memberOf = true)
    @JRuleTag({Tags.LIGHT})
    public void lightProxyToLight(JRuleItemEvent event) {
        if (JRuleItem.forName(event.getMemberName()).isGroup()) {
            return;
        }
        var lightName = event.getMemberName().substring(12);
        logInfo("forwarding proxy to real item");
        var itemLight = JRuleItem.forName(String.format("Light_%s", lightName));
        itemLight.sendCommand(event.getState().getValueAsInteger());
    }
```
I don't know what kind of Item Light_XXX is (could be Switch or Dimmer or Color)
The option would be to check the instance of itemLight, but then I have to check the instance of event.getState() as well.

querdenker2k commented:
I changed the class of the state property of the JRuleItemEvent to JRuleValue now. Shall I do this in this pull request as well?
@seime @seaside1

querdenker2k commented:
push

seaside1 commented:
Yes you might as well I recon (JRuleItemEvent / JRuleValue). I'm running testing this in my prod instance now. I'll report back soon.
How is it going for you @seime ? It seems my rules are not affected to any larger extent.
Regarding naming, I guess we have JRuleItems, so JRuleItemNames would follow that pattern.

seime commented:
I've been unable to do any review yet, hopefully will find some time within a week

seaside1 commented:
I'm running it without issue.
- Maybe we should add back support for RGB for ColorItems, that could be done at a later stage
-  Some of the the changes might be a bit subjective if they are actually making things clearer or better, but I don't see any issues with it right now and overall it's an improvement. 
- I like the added tests.
I think we should merge this one first before we start merging the other PR:s, Since it such a big PR I think it's good if Seime can give a review as well.

seime reviewed (commented):
I've added a few preliminary comments. Lots of code changes, will take many more hours to through it all. Findings so far:

- Some missing sendCommand/postUpdates
- A bit inconsistent when JRuleXXXItem has constants for some of the textual representation of enum values and when not
- We should introduce JRuleItem.postUndef() to set items to UNDEF
- Similar JRuleItem.refresh() to send command `RefreshType`
- `sendCommand(String)` and `postUpdate(String)` should be reserved for `String` items only, and the existing should be renamed to `sendUncheckedCommand` to signal to the user that he/she basically overrides type checking and that the consequence is blowup in production
- Some confusing sendCommand(boolean) around
- We should add Javadoc to user facing methods
- Could we merge `JRuleItems` and `JRuleItemNames` by adding an inner static class to `JRuleItems`? We could then refer to item names like `JRuleItems.Name.MyItemName`

Awaiting some response from @querdenker2k  and @seaside1  before continuing code review

querdenker2k commented:
> I've added a few preliminary comments. Lots of code changes, will take many more hours to through it all. Findings so far:
> 
> * Some missing `sendCommand`/ `postUpdate`
> * A bit inconsistent when `JRuleXXXItem` has constants for some of the textual representation of enum values and when not
> * We should introduce `JRuleItem.postUndef()` to set items to `UNDEF`
> * Similar `JRuleItem.refresh()` to send command `RefreshType`
> * `sendCommand(String)` and `postUpdate(String)` should be reserved for `String` items only, and the existing should be renamed to `sendUncheckedCommand` to signal to the user that he/she basically overrides type checking and that the consequence is blowup in production
> * Some confusing sendCommand(boolean) around
> * We should add Javadoc to user facing methods
> * Could we merge `JRuleItems` and `JRuleItemNames` by adding an inner static class to `JRuleItems`? We could then refer to item names like `JRuleItems.Name.MyItemName`
> 
> Awaiting some response from @querdenker2k and @seaside1 before continuing code review

1. Which sendCommand/postUpdate method you mean? I am using this MR for 3 weeks now
2. I will add the missing ones i find.
3. This is realized via postUpdate(null). I had another branch which cleaned up the value stuff, merged this into this branch now
4. I saw this type in the oh code but didn't when to use it so i didn't implemented it. Is it realy used?
5. I don't understand how the sendCommand() with String parameter worked before because in the oh Items Stringtype is not registered as supported command/state.
6. I added the boolean methods to do something like sendCommand(state.equals("foo")) -> true is always on/open/up/100%
7. I will add more javadoc
8. I would not like to merge this because the generated class is getting bigger and bigger. Currently I have ~2700 Items.

querdenker2k commented on src/main/java/org/openhab/automation/jrule/rules/value/JRuleDateTimeValue.java:
```diff
@@ -0,0 +1,104 @@
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRuleDateTimeValue implements JRuleValue {
+    private static final DateTimeFormatter PARSER = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]");
+    private static final DateTimeFormatter PARSER_TZ = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]z");
+    private static final DateTimeFormatter PARSER_TZ_RFC = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]Z");
+    private static final DateTimeFormatter PARSER_TZ_ISO = DateTimeFormatter.ofPattern(
+            "yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS][.SSSSSSSS][.SSSSSSS][.SSSSSS][.SSSSS][.SSSS][.SSS][.SS][.S]]X");
+    private static final DateTimeFormatter FORMATTER_TZ_RFC = DateTimeFormatter
+            .ofPattern("yyyy-MM-dd'T'HH:mm[:ss[.SSSSSSSSS]]Z");
+
```
This is removed with the merge (of my other branch). The JRule values will just wrap up the oh types

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactItem.java:
```diff
@@ -0,0 +1,37 @@
+        super(name, label, type, id);
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), JRuleOpenClosedValue.valueOf(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), JRuleOpenClosedValue.valueOf(command));
+    }
```
JRuleOpenClosedValue is the default type, see the generic type and JRuleItem

querdenker2k replied:
generics has been removed again

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalDimmerItem.java:
```diff
@@ -0,0 +1,59 @@
+        JRuleEventHandler.get().postUpdate(name, new JRulePercentValue(value));
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), new JRulePercentValue(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), new JRulePercentValue(command));
+    }
```
I would like to do this sendCommand(value.equals("turn it on"));

seime replied:
Can't you do that with `sendCommand(value.equals("turn it on") ? 100 : 0)`?

At least it should be added javadoc to the method to explain how it actually works

querdenker2k replied:
of course, but its longer ;)
javadoc is added

querdenker2k replied:
removed boolean options for all item types, except SwitchItem

querdenker2k commented on src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:
```diff
@@ -0,0 +1,50 @@
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRulePercentValue extends JRuleDecimalValue {
+    public JRulePercentValue(int value) {
+        super(value);
+    }
+
+    public JRulePercentValue(double value) {
+        super((int) Math.round(value + 0.5));
+    }
+
```
within the merged branch this is handled by the oh type

querdenker2k commented on src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:
```diff
@@ -0,0 +1,50 @@
+        super(value);
+    }
+
+    public JRulePercentValue(boolean value) {
+        super(value ? 100 : 0);
+    }
+
```
as above suggested why not ;)

querdenker2k replied:
removed

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalImageItem.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author Arne Seime - Initial contribution
+ */
+public class JRuleInternalImageItem extends JRuleInternalItem<JRuleRawValue> implements JRuleImageItem {
```
JRuleRawValue is the only accepted type

seime replied:
But the default impl also adds a sendCommand which is not supported by openhab?

querdenker2k replied:
thats right for now, actually all items allow all JRuleValues for easier usage (for me). the other option would be to again allow just the supported ones and add a "sendUnsafeCommand" and "postUnsafeUpdate" which accept all JRuleValues.

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalLocationItem.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @author Arne Seime - Initial contribution
+ */
+public class JRuleInternalLocationItem extends JRuleInternalItem<JRulePointValue> implements JRuleLocationItem {
```
see JRuleItem. its supported

seime replied:
You're rigth

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalNumberItem.java:
```diff
@@ -0,0 +1,86 @@
+ *
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRuleInternalNumberItem extends JRuleInternalItem<JRuleDecimalValue> implements JRuleNumberItem {
```
i hope this is handled by the wrapped oh type

querdenker2k replied:
enhancing the QuantityTypes will be a new merge-request

querdenker2k commented on src/main/java/org/openhab/automation/jrule/items/JRuleItem.java:
```diff
@@ -17,59 +17,82 @@
+    }
 
-    public abstract String getType();
+    default void sendCommand(String state) {
+        JRuleEventHandler.get().sendCommand(getName(), new JRuleStringValue(state));
+    }
```
you're right, was already removed in the other branch

querdenker2k commented on src/main/java/org/openhab/automation/jrule/items/JRuleItem.java:
```diff
@@ -17,59 +17,82 @@
+    }
 
-    public abstract String getId();
+    default void postUpdate(String state) {
+        JRuleEventHandler.get().postUpdate(getName(), new JRuleStringValue(state));
+    }
```
see above

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java:
```diff
@@ -0,0 +1,45 @@
+        JRuleEventHandler.get().postUpdate(name, state);
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), JRulePlayPauseValue.valueOf(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), JRulePlayPauseValue.valueOf(command));
+    }
```
see above

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalPlayerItem.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRuleInternalPlayerItem extends JRuleInternalItem<JRulePlayPauseValue> implements JRulePlayerItem {
```
i allowed currently JRuleValue (in the merged branch)

querdenker2k replied:
added

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalRollershutterItem.java:
```diff
@@ -0,0 +1,60 @@
+        JRuleEventHandler.get().postUpdate(name, new JRulePercentValue(value));
+    }
+
+    public void sendCommand(boolean command) {
+        JRuleEventHandler.get().sendCommand(getName(), JRuleUpDownValue.valueOf(command));
+    }
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().postUpdate(getName(), JRuleUpDownValue.valueOf(command));
+    }
```
i will add javadoc for this

querdenker2k replied:
removed boolean option

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalStringItem.java:
```diff
@@ -10,26 +10,19 @@
-    public JRulePercentType(double value) {
-        this.value = (int) Math.round(value + 0.5);
-    }
+public class JRuleInternalStringItem extends JRuleInternalItem<JRuleStringValue> implements JRuleStringItem {
```
changed in merged branch

querdenker2k commented on src/main/java/org/openhab/automation/jrule/items/JRuleContactItem.java:
```diff
@@ -12,36 +12,29 @@
-        super(itemName);
-    }
+public interface JRuleContactItem extends JRuleItem<JRuleOpenClosedValue> {
+    String OPEN = "OPEN";
+    String CLOSED = "CLOSED";
 
```
you're right, will change this

querdenker2k commented on src/main/java/org/openhab/automation/jrule/items/JRulePlayerItem.java:
```diff
@@ -12,44 +12,29 @@
-        super(itemName);
-    }
+public interface JRulePlayerItem extends JRuleItem<JRulePlayPauseValue> {
+    String PLAY = "PLAY";
+    String PAUSE = "PAUSE";
```
will change this.

seime replied:
Missing sendCommand/postUpdate for the mentioned types

querdenker2k replied:
added

querdenker2k commented:
will add javadocs in the evening.

seaside1 commented:
What is the status now then?
Yes I agree that it can be a bit of fuzzy logic to have a for instance. JRuleContactItem.sendCommand(boolean value)
In that case I think it's better to use JRuleOpenCloseValue or similar.

querdenker2k commented:
on my behalf, i will remove all sendCommand/postUpdate with boolean parameters.
what else?

seime commented:
Values are now wrapped in a JRuleValue-class

seime commented on src/main/java/org/openhab/automation/jrule/rules/value/JRulePercentValue.java:
```diff
@@ -0,0 +1,91 @@
+ *
+ * @author Joseph (Seaside) Hagberg - Initial contribution
+ */
+public class JRulePercentValue implements JRuleValue {
```
Should this not extend JRuleDecimalValue to align with openhab internals?

querdenker2k replied:
done

seime reviewed (commented):
sendCommand(boolean) and postUpdate(boolean) methods seems to be still present around, I think you wrote you'd remove them?

seime commented:
It is not possible do a `postUpdate(null)` whenever there overloaded versions. Must introduce a fake cast to one of the methods ie `postUpdate((Date) null)` to achieve this

seime commented:
Question: Ie getState().asStringValue() when state is null. Will getState() return null, or getState().asStringValue() (and others) return null?

seime commented:
I also find it a bit painful to do runtime casts of ie JRuleItemEvent.getState()/getOldState(). Blows up at runtime / unit tests if not correct. Could we find a more typesafe solution?

seime commented:
Should we have a separate JRuleQuantityTypeItem / internal class to avoid mixing these together? Now it is possible to send a quantity/unit to a non-quantity type item?

querdenker2k commented:
> sendCommand(boolean) and postUpdate(boolean) methods seems to be still present around, I think you wrote you'd remove them?

yes, i said i will do. but i don't want to do every small change "right now".
done now

querdenker2k commented:
> It is not possible do a `postUpdate(null)` whenever there overloaded versions. Must introduce a fake cast to one of the methods ie `postUpdate((Date) null)` to achieve this

added extra method for this **postNullUpdate**

querdenker2k commented:
> Question: Ie getState().asStringValue() when state is null. Will getState() return null, or getState().asStringValue() (and others) return null? If the first this now requires the rule code to be littered with null checks for getState().

done

querdenker2k commented:
> I also find it a bit painful to do runtime casts of ie JRuleItemEvent.getState()/getOldState(). Blows up at runtime / unit tests if not correct. Could we find a more typesafe solution?

maybe we can do this later, this PR is getting bigger and bigger...

querdenker2k commented:
> Should we have a separate JRuleQuantityTypeItem / internal class to avoid mixing these together? Now it is possible to send a quantity/unit to a non-quantity type item?

probably we should, but i would say in another PR.

seime commented:
OK, could you just resolve the open conversations that indeed have been resolved, and keep the unresolved open to make it easy to create separate tasks?

Fully agreed that the PR is big and should be merged now. Great wort @querdenker2k !

Do you want to take a final look and merge if you're happy @seaside1 ?

querdenker2k commented on src/main/java/org/openhab/automation/jrule/internal/items/JRuleInternalContactGroupItem.java:
```diff
@@ -10,48 +10,43 @@
+
+    public void postUpdate(boolean command) {
+        JRuleEventHandler.get().getGroupMemberNames(name, false)
+                .forEach(s -> JRuleEventHandler.get().postUpdate(s, JRuleOnOffValue.valueOf(command)));
```
resolved as above

querdenker2k commented on src/main/java/org/openhab/automation/jrule/items/JRuleItem.java:
```diff
@@ -17,59 +17,82 @@
 
-    public String getName() {
-        return itemName;
+    default void sendCommand(T command) {
```
done

seaside1 commented:
Yes looks good. Lets merge!

seaside1 reviewed (approved):
Approved!
